### PR TITLE
Upgrade to Scala `2.13.12` 

### PR DESCRIPTION
**Pull request by ioannakok**
## What does this change?
Upgrades scala version to `2.13.12`.

## What is the value of this and can you measure success?
* Keeping up-to-date with the latest Scala 2.13 version
* Addresses "Implicit definition should have explicit type" warnings which in our case become errors because we have `-Xfatal-warnings` flag enabled. This also helps to ease migration to Scala 3.
* Unblocks:
  * https://github.com/guardian/frontend/pull/26783 
With Scala `2.13.10` the compiler was crashing.

## Screenshots

| Before      | After      |
|-------------|------------|
| image | image |

Instead of crashing the compiler is producing a compilation error which we can now work on fixing in #26783 

## Checklist

- [x] Tested locally, and on CODE if necessary
- [x] Will not break dotcom-rendering
- [x] Will not break our database – if updating CAPI, [updated and committed the database files](https://github.com/guardian/frontend/blob/main/docs/03-dev-howtos/15-updating-test-database.md)

### CODE REVIEW

File: common/app/common/RequestLogger.scala [hunk 1]
```diff
@@ -31,7 +31,7 @@ case class RequestLoggerFields(request: RequestHeader, response: Option[Result],
 
     val allowListedHeaders = (for {
       headerName <- allowListedHeaderNames
-      value <- allHeadersFields.get(headerName)
+      value <- allHeadersFields.map { case (key, value) => (key.toLowerCase, value) }.get(headerName.toLowerCase)
```
Comment (ioannakok) [inline common/app/common/RequestLogger.scala:34]: It looks like in Scala 2.13.12 the `get` method of a `Map` becomes case sensitive, so I had to add this to fix a failing test.

Comment (mkurz) [reply common/app/common/RequestLogger.scala:34]: I was looking into that because this seemed a bit strange to me, to have that code failing solely by upgrading to Scala 2.13.12. (Sorry for the long comment, I just wanted to know myself what was going on here)

**TLDR**: Turns out this is _not_ a bug in Scala and also they did not really change _expected_ behaviour (tough they did change some internals which causes your code to fail). The actual problem with this code was that you assumed that map.get here is case insensitive, which isn't defined somewhere and you can not expect that.

If you look at the code:
https://github.com/guardian/frontend/blob/c683b15c3f1371154220105af0c39420e7d7f2fd/common/app/common/RequestLogger.scala#L30-L35
`allHeadersFields` is a `Map`, not a `SortedMap` or anything concrete, but a "stupid" `Map`. Calling `map.get` usually checks for equality to retrieve an entry. Since you do not know which concrete `Map` implemention is used to back `allHeadersFields`, you can not assume that retrieving elements is case sensitive, since you don't know how elements are compared internally, the `Map` does not guarantee anything here.

The reason your code worked before was pure coincidence. Calling `toMap`
```scala
request.headers.toMap
```
here again just gives you a "stupid" map. Turns out, in Play this gives you a `TreeMap` implemention with a `CaseInsensitiveOrdered` ordering, see [here](https://github.com/playframework/playframework/blob/2.9.1/core/play/src/main/scala/play/api/mvc/Headers.scala#L90):
```scala
val builder = TreeMap.newBuilder[String, Seq[String]](CaseInsensitiveOrdered)
```

Now, calling `mapV` on this method "transforms" this `map`, you can see that [here](https://github.com/rtyley/scala-collection-plus/blob/v0.11/collection-plus/src/main/scala/com/madgag/scala/collection/decorators/internal/MapDecorator.scala#L12):
```scala
  def mapV[U](f: map.V => U): Map[map.K, U] = map(coll).toMap.transform { case (_,v) => f(v) }
```

The `toMap` call here (which basically is `toMap` on the given map) in that line is calling Scala's internal code which in Scala 2.13.11 would return just the same instance that was given already, meaning it just returns the exact same map with the internal `CaseInsensitiveOrdered` ordering.
In Scala 2.13.12 however they changed some internals (because of other reasons) you get back a new instance of `Map` which however does not have the same ordering anymore. So the case-insensitive ordering from Play's internal code was lost and that is why your code breaks. Actually this is totally ok, because `toMap` does not make any promises which map it returns, not does it make promises on ordering.

The internal code they changed in Scala can be seen [here](https://github.com/scala/scala/pull/10337/files#diff-117dce8d6af394ff848de94139ccb7796b1c0ffe8e4f4b872414ce0f1cb2c5f5R210), it's the `from` method: 
In Scala 2.13.11 the matching case was:
```scala
case m: Map[K, V] => m
```
which just returns the same instance again (which was the same `TreeMap` with the ordering)
In Scala 2.13.12 however that case was removed and the code falls back to:
```scala
case _ => newBuilder[K, V].addAll(it).result()
```
which creates a new map, but as you can see the `newBuilder` does not get an `ordering` passed (turns out it then returns an `Map2`).

Comment (ioannakok) [reply common/app/common/RequestLogger.scala:34]: Thanks for the comment! This makes sense

File: project/ProjectSettings.scala [hunk 2]
```diff
@@ -32,7 +32,7 @@ object ProjectSettings {
     Compile / packageDoc / publishArtifact := false,
     Compile / doc / sources := Seq.empty,
     Compile / doc := target.map(_ / "none").value,
-    scalaVersion := "2.13.10",
+    scalaVersion := "2.13.12",
```
Comment (mxdvl) [inline project/ProjectSettings.scala:35]: This looks fine!

**Review by mxdvl — APPROVED** [review]

File: common/app/common/RequestLogger.scala (hunk 1 above)
Comment (ioannakok) [reply common/app/common/RequestLogger.scala:34]: cc @rtyley I think you will be interested in this!
